application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/enforcer/DefaultLicensingSecurityCacheRuleInvalidator.java

```diff
@@ -94,25 +100,46 @@ public void invalidate(XarInstalledExtension extension)
     {
         boolean locked = acquireLock();
         try {
-            for (String namespace : extension.getNamespaces()) {
-                if (namespace.startsWith("wiki:")) {
-                    WikiReference wikiRef = new WikiReference(namespace.substring(5));
-                    for (XarEntry entry : extension.getXarPackage().getEntries()) {
-                        securityCache.remove(
-                            securityReferenceFactory.newEntityReference(
-                                documentReferenceResolver.resolve(entry, wikiRef)
-                            )
-                        );
+            Collection<String> namespaces = extension.getNamespaces();
+            if (namespaces != null) {
+                for (String namespace : namespaces) {
+                    if (namespace.startsWith("wiki:")) {
+                        invalidateForWiki(extension, new WikiReference(namespace.substring(5)));
                     }
                 }
+            } else {
+                // As the extension is installed at farm level, we have to invalidate every cache entry for this
+                // extension in every wiki.
+                for (String wikiId : wikiDescriptorManager.getAllIds()) {
+                    invalidateForWiki(extension, new WikiReference(wikiId));
+                }
             }
+        } catch (WikiManagerException e) {
+            // Fail silently
```

Comment:
You should never fail silently when an exception is thrown. It clearly signals a problem and it should be handled.

Normally, the exception should be thrown up the execution chain, however, the current method does not throw an exception. It would be recommended that it does or you could even wrap it into an unchecked IllegalStateException and still throw it. Worst case, you should at least log the exception (as an error).

Reply:
Logging the error would be better IMO